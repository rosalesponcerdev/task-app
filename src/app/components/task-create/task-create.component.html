<form class="row" [formGroup]="taskCreatePresenter.noteForm" (ngSubmit)="taskCreatePresenter.sendFormHandler()">
  <div class="col-6">
    <div class="row">
      <div class="col">
        <div class="row align-items-center">
          <div class="col-auto">
            <label>Tarea:</label>
          </div>
          <div class="col">
            <input formControlName="title" type="text" class="form-control" maxlength="40" />
          </div>
        </div>
      </div>
      <div class="col-auto">
        <button [disabled]="!taskCreatePresenter.noteForm.valid" type="submit" class="btn btn-primary">Agregar</button>
      </div>
    </div>
  </div>
</form>
