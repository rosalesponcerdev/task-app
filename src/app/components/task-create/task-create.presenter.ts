import { Injectable } from '@angular/core';
import { FormControl, FormGroup, Validators } from '@angular/forms';
import { Task } from '../../interface/task.interface';
import { Observable, Subject, takeUntil } from 'rxjs';

@Injectable()
export class TaskCreatePresenter {
  noteForm: FormGroup<{
    title: FormControl<string>;
  }>;

  sendFormHandler$: Observable<Task>;
  private _sendFormHandler: Subject<Task>;

  private _destroy = new Subject();

  constructor() {
    this._sendFormHandler = new Subject<Task>();
    this.sendFormHandler$ = this._sendFormHandler.asObservable().pipe(takeUntil(this._destroy));

    this.noteForm = this._buildNoteForm();
  }

  ngOnDestroy(): void {
    this._destroy.next(true);
    this._destroy.complete();
  }

  getFormValue() {
    return this.noteForm.value as Task;
  }

  sendFormHandler() {
    if (!this.noteForm.valid) return;

    this._sendFormHandler.next(this.getFormValue());

    this.noteForm.reset();
  }

  private _buildNoteForm() {
    return new FormGroup({
      title: new FormControl('', {
        nonNullable: true,
        validators: [Validators.required, Validators.minLength(1)],
      }),
    });
  }
}
