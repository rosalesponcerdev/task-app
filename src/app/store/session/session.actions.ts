import { createAction, props } from '@ngrx/store';
import { Session } from '../../interface/session.interface';
import { LoginForm } from '../../interface/login-form.interface';

export const login = createAction('[Session] Login', props<LoginForm>());
export const loginSuccess = createAction('[Session] Login Success', props<{ session: Session }>());
export const loginFailure = createAction('[Session] Login Failure', props<{ error: any }>());
export const setLoading = createAction('[Session] Set Loading', props<{ isLoading: boolean }>());

export const logout = createAction('[Session] Logout');
export const logoutSuccess = createAction('[Session] logout Success');
