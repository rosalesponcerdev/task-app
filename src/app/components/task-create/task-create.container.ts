import { Component } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Store } from '@ngrx/store';
import { Observable } from 'rxjs';

import { selectCompleteTaskSList } from '../../store/task/task.selectors';
import { addTask } from '../../store/task/task.actions';

import { Task } from '../../interface/task.interface';

import { TaskCreateComponent } from './task-create.component';

@Component({
  selector: 'task-create-container',
  standalone: true,
  imports: [CommonModule, TaskCreateComponent],
  template: `
    <app-task-create-ui (sendForm)="sendFormHandler($event)" />
  `,
})
export class TaskCreateContainer {
  tasks$: Observable<Task[]>;

  constructor(private _store: Store) {
    this.tasks$ = this._store.select(selectCompleteTaskSList);
  }

  ngOnInit() {}

  sendFormHandler(task: Task) {
    task.id = crypto.randomUUID();

    this._store.dispatch(addTask({ task: task }));
  }
}
