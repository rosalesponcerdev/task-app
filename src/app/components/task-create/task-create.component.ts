import { CommonModule } from '@angular/common';
import { Component, EventEmitter, Output } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';

import { Observable } from 'rxjs';

import { Store } from '@ngrx/store';
import { TaskCreatePresenter } from './task-create.presenter';
import { selectName } from '../../store/session/session.selectors';
import { Task } from '../../interface/task.interface';

@Component({
  selector: 'app-task-create-ui',
  standalone: true,
  imports: [CommonModule, ReactiveFormsModule, FormsModule],
  templateUrl: './task-create.component.html',
  providers: [TaskCreatePresenter],
})
export class TaskCreateComponent {
  @Output() sendForm = new EventEmitter<Task>();

  username$: Observable<string | null>;

  constructor(private _store: Store, public taskCreatePresenter: TaskCreatePresenter) {
    this.username$ = this._store.select(selectName);

    this.taskCreatePresenter.sendFormHandler$.subscribe((formValue) => {
      this.sendForm.emit(formValue);
    });
  }
}
