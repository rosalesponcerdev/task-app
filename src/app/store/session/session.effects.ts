// src/app/state/session/session.effects.ts
import { Injectable } from '@angular/core';

import { merge, of } from 'rxjs';
import { catchError, map, mergeMap, tap } from 'rxjs/operators';

import * as SessionActions from './session.actions';
import * as TaskActions from '../task/task.actions';
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { AuthService } from '../../services/auth.service';
import { Router } from '@angular/router';
import { Store } from '@ngrx/store';

@Injectable()
export class SessionEffects {
  login$ = createEffect(() =>
    this._actions.pipe(
      ofType(SessionActions.login),
      tap(() => this._store.dispatch(SessionActions.setLoading({ isLoading: true }))),
      mergeMap((action) =>
        this._authService.login(action.username, action.password).pipe(
          tap(() => this._store.dispatch(SessionActions.setLoading({ isLoading: false }))),
          mergeMap((session) => [SessionActions.loginFailure({ error: null }), SessionActions.loginSuccess({ session })]),
          catchError((error) => of(SessionActions.loginFailure({ error: error })))
        )
      )
    )
  );

  loginSuccess$ = createEffect(
    () => {
      return this._actions.pipe(
        ofType(SessionActions.loginSuccess),
        tap(() => this._router.navigate(['/']))
      );
    },
    {
      dispatch: false,
    }
  );

  logout$ = createEffect(() =>
    this._actions.pipe(
      ofType(SessionActions.logout),
      mergeMap(() =>
        this._authService.logout().pipe(
          mergeMap(() => [TaskActions.clearAllTasks(), SessionActions.logoutSuccess()]),
          catchError((error) => of(SessionActions.loginFailure({ error })))
        )
      )
    )
  );

  logoutSuccess$ = createEffect(
    () => {
      return this._actions.pipe(
        ofType(SessionActions.logoutSuccess),
        tap(() => {
          this._router.navigate(['/login']);
        })
      );
    },
    {
      dispatch: false,
    }
  );

  constructor(private _actions: Actions, private _authService: AuthService, private _router: Router, private _store: Store) {}
}
