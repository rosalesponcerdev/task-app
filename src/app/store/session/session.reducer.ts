import { createReducer, on } from '@ngrx/store';

import { loginFailure, loginSuccess, logout, setLoading } from './session.actions';
import { Session } from '../../interface/session.interface';

export interface SessionState {
  session: Session | null;
  error: any;
  isLoading: boolean;
  name: string | null;
}

export const initialState: SessionState = {
  session: null,
  error: null,
  isLoading: false,
  name: null,
};

export const sessionReducer = createReducer(
  initialState,

  on(loginSuccess, (state, { session }) => {
    let name = '';

    try {
      name = JSON.parse(atob(session.token.split('.')[1])).name;
    } catch (error) {}

    return {
      ...state,
      session,
      error: null,
      isLoading: false,
      name,
    };
  }),
  on(loginFailure, (state, { error }) => ({
    ...state,
    session: null,
    error,
    isLoading: false,
  })),
  on(logout, (state) => ({
    ...state,
    session: null,
    error: null,
    isLoading: false,
  })),
  on(setLoading, (state, { isLoading }) => ({
    ...state,
    isLoading,
  }))
);
