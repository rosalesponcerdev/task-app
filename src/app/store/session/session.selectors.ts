import { createFeatureSelector, createSelector } from '@ngrx/store';

import { SessionState } from './session.reducer';

export const selectSessionState = createFeatureSelector<SessionState>('session');

export const selectSessionError = createSelector(selectSessionState, (state: SessionState) => state.error);
export const selectIsLoggedIn = createSelector(selectSessionState, (state: SessionState) => !!state.session);
export const selectIsLoading = createSelector(selectSessionState, (state: SessionState) => !!state.isLoading);
export const selectName = createSelector(selectSessionState, (state: SessionState) => state.name);
