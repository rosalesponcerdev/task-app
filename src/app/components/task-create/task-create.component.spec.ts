import { ComponentFixture, TestBed } from '@angular/core/testing';

import { TaskCreateComponent } from './task-create.component';
import { provideStore } from '@ngrx/store';
import { sessionReducer } from '../../store/session/session.reducer';
import { taskReducer } from '../../store/task/task.reducer';
import { provideEffects } from '@ngrx/effects';
import { SessionEffects } from '../../store/session/session.effects';

describe('TaskCreateComponent', () => {
  let component: TaskCreateComponent;
  let fixture: ComponentFixture<TaskCreateComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [TaskCreateComponent],
      providers: [
        provideStore({
          session: sessionReducer,
          task: taskReducer,
        }),
        provideEffects(SessionEffects),
      ],
    }).compileComponents();

    fixture = TestBed.createComponent(TaskCreateComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('Happy Path', () => {
    expect(component).toBeTruthy();

    expect(fixture.nativeElement).toMatchSnapshot();

    component.taskCreatePresenter.noteForm.patchValue({
      title: 'title',
    });

    component.taskCreatePresenter.noteForm.updateValueAndValidity();

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
    expect(component.taskCreatePresenter.noteForm.valid).toBeTruthy();

    component.taskCreatePresenter.sendFormHandler();
  });

  it('Happy Error', () => {
    expect(component).toBeTruthy();

    expect(fixture.nativeElement).toMatchSnapshot();

    component.taskCreatePresenter.noteForm.patchValue({
      title: '',
    });

    component.taskCreatePresenter.noteForm.updateValueAndValidity();

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
    expect(component.taskCreatePresenter.noteForm.valid).toBeFalsy();

    component.taskCreatePresenter.sendFormHandler();
  });
});
