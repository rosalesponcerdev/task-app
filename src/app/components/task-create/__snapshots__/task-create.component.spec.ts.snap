// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskCreateComponent Happy Error 1`] = `
<div
  id="root1"
>
  <form
    class="row ng-untouched ng-pristine ng-invalid"
    novalidate=""
  >
    <div
      class="col-6"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          <div
            class="row align-items-center"
          >
            <div
              class="col-auto"
            >
              <label>
                Tarea:
              </label>
            </div>
            <div
              class="col"
            >
              <input
                class="form-control ng-untouched ng-pristine ng-invalid"
                formcontrolname="title"
                maxlength="40"
                type="text"
              />
            </div>
          </div>
        </div>
        <div
          class="col-auto"
        >
          <button
            class="btn btn-primary"
            disabled=""
            type="submit"
          >
            Agregar
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`TaskCreateComponent Happy Error 2`] = `
<div
  id="root1"
>
  <form
    class="row ng-untouched ng-pristine ng-invalid"
    novalidate=""
  >
    <div
      class="col-6"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          <div
            class="row align-items-center"
          >
            <div
              class="col-auto"
            >
              <label>
                Tarea:
              </label>
            </div>
            <div
              class="col"
            >
              <input
                class="form-control ng-untouched ng-pristine ng-invalid"
                formcontrolname="title"
                maxlength="40"
                type="text"
              />
            </div>
          </div>
        </div>
        <div
          class="col-auto"
        >
          <button
            class="btn btn-primary"
            disabled=""
            type="submit"
          >
            Agregar
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`TaskCreateComponent Happy Path 1`] = `
<div
  id="root0"
>
  <form
    class="row ng-untouched ng-pristine ng-invalid"
    novalidate=""
  >
    <div
      class="col-6"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          <div
            class="row align-items-center"
          >
            <div
              class="col-auto"
            >
              <label>
                Tarea:
              </label>
            </div>
            <div
              class="col"
            >
              <input
                class="form-control ng-untouched ng-pristine ng-invalid"
                formcontrolname="title"
                maxlength="40"
                type="text"
              />
            </div>
          </div>
        </div>
        <div
          class="col-auto"
        >
          <button
            class="btn btn-primary"
            disabled=""
            type="submit"
          >
            Agregar
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`TaskCreateComponent Happy Path 2`] = `
<div
  id="root0"
>
  <form
    class="row ng-untouched ng-pristine ng-valid"
    novalidate=""
  >
    <div
      class="col-6"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          <div
            class="row align-items-center"
          >
            <div
              class="col-auto"
            >
              <label>
                Tarea:
              </label>
            </div>
            <div
              class="col"
            >
              <input
                class="form-control ng-untouched ng-pristine ng-valid"
                formcontrolname="title"
                maxlength="40"
                type="text"
              />
            </div>
          </div>
        </div>
        <div
          class="col-auto"
        >
          <button
            class="btn btn-primary"
            type="submit"
          >
            Agregar
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;
